crates/starknet-devnet-server/src/api/json_rpc/mod.rs

```diff
@@ -77,128 +80,155 @@ impl ToRpcResponseResult for StrictRpcResult {
                 declare_transaction,
             }) => {
                 let BroadcastedDeclareTransactionEnumWrapper::Declare(broadcasted_transaction) =
                     declare_transaction;
-                self.add_declare_transaction(broadcasted_transaction).await.to_rpc_result()
+                self.add_declare_transaction(broadcasted_transaction).await
             }
             StarknetRequest::AddDeployAccountTransaction(
                 BroadcastedDeployAccountTransactionInput { deploy_account_transaction },
             ) => {
+                forwardable = false;
                 let BroadcastedDeployAccountTransactionEnumWrapper::DeployAccount(
                     broadcasted_transaction,
                 ) = deploy_account_transaction;
-                self.add_deploy_account_transaction(broadcasted_transaction).await.to_rpc_result()
+                self.add_deploy_account_transaction(broadcasted_transaction).await
             }
             StarknetRequest::AddInvokeTransaction(BroadcastedInvokeTransactionInput {
                 invoke_transaction,
             }) => {
                 let BroadcastedInvokeTransactionEnumWrapper::Invoke(broadcasted_transaction) =
                     invoke_transaction;
-                self.add_invoke_transaction(broadcasted_transaction).await.to_rpc_result()
+                self.add_invoke_transaction(broadcasted_transaction).await
+            }
+            StarknetRequest::EstimateMessageFee(request) => {
+                self.estimate_message_fee(request.get_block_id(), request.get_raw_message().clone())
+                    .await
             }
-            StarknetRequest::EstimateMessageFee(request) => self
-                .estimate_message_fee(request.get_block_id(), request.get_raw_message().clone())
-                .await
-                .to_rpc_result(),
             StarknetRequest::SimulateTransactions(SimulateTransactionsInput {
                 block_id,
                 transactions,
                 simulation_flags,
-            }) => self
-                .simulate_transactions(block_id, transactions, simulation_flags)
-                .await
-                .to_rpc_result(),
+            }) => self.simulate_transactions(block_id, transactions, simulation_flags).await,
             StarknetRequest::TraceTransaction(TransactionHashInput { transaction_hash }) => {
-                self.get_trace_transaction(transaction_hash).await.to_rpc_result()
+                self.get_trace_transaction(transaction_hash).await
             }
             StarknetRequest::BlockTransactionTraces(BlockIdInput { block_id }) => {
-                self.get_trace_block_transactions(block_id).await.to_rpc_result()
+                self.get_trace_block_transactions(block_id).await
+            }
+        };
+
+        if let (Err(err), Some(forwarder)) = (&starknet_resp, &self.origin_caller) {
+            match err {
```

Comment:
Is this the best approach to forward requests to forked network? What are the guarantees that some of the other methods in the chain won't return 1 of these 4 errors and a method that shouldn't be forwarded is forwarded. Also the other way around, what will happen If some of the methods (candidates to fetch data from forked network) change their error variants?

Reply:
I inspected the RPC specification and the code, and my conclusion is that this should work. As for future changes, we would somehow need to remember to check for new errors in future spec versions - perhaps we can add that as a note to the `crates/starknet-devnet-server/test_data/spec/0.7.0/edit_spec_instructions.yaml`?

Reply:
this yml is not good candidate for such notes, because it is only for modifying the spec in the tests to not generate data that can't be handled by the application

Reply:
In README.md we have some `## Development` sections, perhaps a `## Development - RPC specification update` could be added, though no one guaranteers that future updaters will give it a look.

Reply:
yes indeed relying on documentation is not a guarantee

Reply:
We could add another checkbox entry to the PR template that reminds the developer to go through the `## Development section`